src/lib/integration_test_lib/malleable_error.ml

```diff
@@ -0,0 +1,345 @@
+ *  extraction. Hard errors effect the control flow of the monad in the same
+ *  way an `Error` constructor for `Or_error.t` would.
+ It remains similar to Deferred.Or_error.t in that it is a specialization of Deferred.Result.t
+ *)
+
+module Hard_fail = struct
+  type t =
+    { hard_error: Test_error.internal_error
+    ; soft_errors: Test_error.internal_error list }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module Accumulator = struct
+  type 'a t =
+    {computation_result: 'a; soft_errors: Test_error.internal_error list}
+  [@@deriving eq, sexp_of, compare]
+end
+
+type 'a t = ('a Accumulator.t, Hard_fail.t) Deferred.Result.t
+
+module T = Monad.Make (struct
+  type nonrec 'a t = 'a t
+
+  let return a =
+    Deferred.return (Ok {Accumulator.computation_result= a; soft_errors= []})
+
+  (* 
+
+('a, 'b) result Deferred.t ->
+f:('c -> ('d, 'e) result Deferred.t) -> 
+('f, 'b) result Deferred.t 
+
+*)
+  let bind res ~f =
+    match%bind res with
+    | Ok {Accumulator.computation_result= prev_result; soft_errors= err_list}
+      -> (
+        let execProg = f prev_result in
+        match%bind execProg with
+        | Ok {Accumulator.computation_result= comp; soft_errors= next_list} ->
+            Deferred.return
+              (Ok
+                 { Accumulator.computation_result= comp
+                 ; Accumulator.soft_errors= List.append err_list next_list })
+        | Error
+            {Hard_fail.hard_error= hard_err; Hard_fail.soft_errors= next_list}
+          ->
+            Deferred.return
+              (Error
+                 { Hard_fail.hard_error= hard_err
+                 ; soft_errors= List.append err_list next_list }) )
+    | Error _ as error ->
+        Deferred.return error
+
+  let map = `Define_using_bind
+end)
+
+include T
+
+let return_without_deferred a =
+  Ok {Accumulator.computation_result= a; soft_errors= []}
+
+let return_unit_without_deferred =
+  Ok {Accumulator.computation_result= (); soft_errors= []}
+
+let ok_unit = return ()
+
+let ok_exn (res : 'a t) : 'a Deferred.t =
+  let open Deferred.Let_syntax in
+  match%bind res with
+  | Ok {Accumulator.computation_result= x; soft_errors= _} ->
+      Deferred.return x
+  | Error {Hard_fail.hard_error= err; Hard_fail.soft_errors= _} ->
+      Error.raise err.error
+
+let error_string message =
+  Deferred.return
+    (Error
+       { Hard_fail.hard_error=
+           Test_error.raw_internal_error (Error.of_string message)
+       ; soft_errors= [] })
+
+let errorf format = Printf.ksprintf error_string format
+
+let soft_error res err =
+  Deferred.return
+    (Ok
+       { Accumulator.computation_result= res
+       ; soft_errors= [Test_error.raw_internal_error err] })
+
+let hard_error err =
+  Deferred.return
+    (Error
+       { Hard_fail.hard_error= Test_error.raw_internal_error err
+       ; soft_errors= [] })
+
+let or_error_to_hard_malleable_error (or_err : 'a Deferred.Or_error.t) : 'a t =
```

Comment:
These function names can be shortened. Since we are in a module called `Malleable_error`, we typically think about the naming of our functions underneath that concept in OCaml. A more OCaml'y name for this would be `of_hard_or_error`, which read well under the context `Malleable_error.of_hard_or_error`.